### PR TITLE
WI21 78 AWS code refactoring and KaaIoT added

### PR DESCRIPTION
**Pull request by marek-szydlowski-wizzdev**
Created cloud interface; refactored and separated AWS related code from non-cloud related one; Added new cloud - KaaIoT

### CODE REVIEW

File: MicroPython/README.md [hunk 1]
```diff
@@ -1,54 +1,69 @@
...
+### To develop in PyCharm:
+
+Open the project MicroPython with PyCharm and mark *src* and *ulib_mocks* directory as Sources Root in Pycharm.
+
+## Cloud service provider: 
+
+### - KAA
+Make sure that your KAA cloud is configured. For more information please go to README in "KAA" directory [here](../KAA/README.md).
+
+By now, you should have for things:
```
Comment (marcin-nawrocki-wizzdev) [inline MicroPython/README.md:45]: four* things

File: MicroPython/src/cloud/AWS_cloud.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+            "Allocated bytes on heap before gc {}".format(gc.mem_alloc()))
+        gc.collect()
+        logging.info(
+            "Allocated bytes on heap after gc {}".format(gc.mem_alloc()))
+        self.configure_data_from_terraform()
+        logging.debug("Authorization request...")
+        jwt_token = self.authorization_request()
+
+        if jwt_token is not "":
+
```
Comment (marcin-nawrocki-wizzdev) [inline MicroPython/src/cloud/AWS_cloud.py:70]: remove this enters

File: MicroPython/src/cloud/KAA_cloud.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+        elif topic == self.publish_error_topic:
+            status_code = msg['statusCode']
+            reason = msg['reasonPhrase']
+            logging.info('Operation failed with error code: {} - reason: {}\n'.format(
+                status_code, reason
+            ))
+        else:
+            logging.info('On topic: {} received msg: {}'.format(topic, msg))
+
+    def device_configuration(self, data: dict) -> int:
```
Comment (marcin-nawrocki-wizzdev) [inline MicroPython/src/cloud/KAA_cloud.py:51]: is this exactly the same method as in AWS cloud?

File: MicroPython/src/cloud/KAA_cloud.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        """
+        if utils.check_if_file_exists(config.KAA_CONFIG_PATH) == 0:
+            raise Exception("Create kaa_config.json file first!")
+
+        with open(config.KAA_CONFIG_PATH, "r", encoding="utf8") as infile:
+            config_dict = ujson.load(infile)
+
+        return config_dict
+
+    def _format_data(self, data: dict) -> dict:
```
Comment (marcin-nawrocki-wizzdev) [inline MicroPython/src/cloud/KAA_cloud.py:129]: why not static?

File: MicroPython/src/cloud/KAA_cloud.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+            topic=self.publish_success_topic, callback=self.receive_message, qos=config.cfg.QOS
+        )
+        result_err_topic = mqtt_communicator.subscribe(
+            topic=self.publish_error_topic, callback=self.receive_message, qos=config.cfg.QOS
+        )
+        if not result_suc_topic or not result_err_topic:
+            logging.error(
+                "Error subscribing to topics with MQTT in publish_data()")
+
+        # TODO: Certificates?
```
Comment (marcin-nawrocki-wizzdev) [inline MicroPython/src/cloud/KAA_cloud.py:158]: ?

File: MicroPython/src/common/utils.py [hunk 6]
```diff
@@ -147,38 +144,40 @@ def read_from_file(file_path: str) -> (bool, str):
...
-    Create new instance od MQTTCommunicator.
-    :return: Instance of MQTTCommunicator.
-    """
-    logging.debug("utils.py/create_MQTT_communicator_from_config()")
-    return MQTTCommunicator(use_AWS=config.cfg.use_aws,
-                            client_id=config.cfg.aws_client_id,
-                            endpoint=config.cfg.aws_endpoint,
-                            port=config.cfg.mqtt_port_ssl,
-                            timeout=config.cfg.mqtt_timeout)
+# TODO: Not needed?
```
Comment (marcin-nawrocki-wizzdev) [inline MicroPython/src/common/utils.py:147]: delete if not used

File: MicroPython/src/controller/main_controller.py [hunk 7]
```diff
@@ -39,13 +44,32 @@ def __init__(self):
...
+
+        elif config.cfg.cloud_provider == Providers.KAA:
+            web_app.setup(
+                get_measurement_hook=self.get_measurement,
+                configure_device_hook=self.cloud_provider.device_configuration,
+                configure_sensor_hook=self.configure_sensor,
+                start_test_data_acquisition=self.start_test_data_acquisition_hook,
+                start_data_acquisition=self.start_data_acquisition_hook,
+                get_status_hook=self.get_status
+            )
```
Comment (marcin-nawrocki-wizzdev) [inline MicroPython/src/controller/main_controller.py:71]: Is this if/elif clause needed?

File: MicroPython/src/controller/main_controller.py [hunk 8]
```diff
@@ -76,6 +100,17 @@ def perform(self) -> None:
             self.lock.release()
             utime.sleep_ms(500)
 
+    def get_cloud_provider(self) -> CloudProvider:
```
Comment (marcin-nawrocki-wizzdev) [inline MicroPython/src/controller/main_controller.py:103]: static?

**Review by marcin-nawrocki-wizzdev — COMMENTED** [review]
Please resolve comments and PR will be approved

File: MicroPython/src/cloud/KAA_cloud.py (hunk 3 above)
Comment (marek-szydlowski-wizzdev) [reply MicroPython/src/cloud/KAA_cloud.py:51]: It's not, as the one in AWS code needs to register aws_thing, and the one in KAA needs to adjust (and load) some config fields related to communication with KaaIoT

File: MicroPython/src/cloud/KAA_cloud.py (hunk 4 above)
Comment (marek-szydlowski-wizzdev) [reply MicroPython/src/cloud/KAA_cloud.py:129]: It's a helper function used for formatting a data in a way expected by Kaa. It's rather a private method and I don't think it should be static as there is no use-case except during publishing the data.

File: MicroPython/src/cloud/KAA_cloud.py (hunk 5 above)
Comment (marek-szydlowski-wizzdev) [reply MicroPython/src/cloud/KAA_cloud.py:158]: We're currently unable to create secure connection through SSL as Kaa does not allow us to create certificates (internal server error during generating of a key) so there are no certificates for now.

File: MicroPython/src/controller/main_controller.py (hunk 7 above)
Comment (marek-szydlowski-wizzdev) [reply MicroPython/src/controller/main_controller.py:71]: AWS one requires aws_hook to be supplied but there is no need for it in KAA cloud. I also think it's a bit cleaner to use if/elif as for each cloud service, the setup can be seen independently.

**Review by marcin-nawrocki-wizzdev — APPROVED** [review]
All conversations resolved. Approve.

**Pull request merged**